torch/quantization/convert_modules.py

```diff
@@ -0,0 +1,222 @@
+from __future__ import absolute_import, division, print_function, unicode_literals
+import torch.nn as nn
+import torch.nn.quantized as nnq
+import torch
+from .observer import *
+import heapq
+
+def _forward_hook(self, input, output):
+    for module in self.children():
+        module.forward(output)
+
+def add_observer(module, qconfig_dict, qconfig_parent=None, prefix=''):
+    r"""Transform a module to a quantized module according to qconfig_dict
+
+    This function insert observer module to all leaf child module of a given
+    module based on qconfig_dict.
+
+    Args:
+        module: instance of the module we want to transform
+        qconfig_dict: dictionary that maps from name of submodule to quantization
+                     configuration
+        qconfig_parent: quantization config of parent module, we will fallback to
+                       this config when there is no specified config for current
+                       module
+        prefix: corresponding prefix of the current module, used as key in
+                qconfig_dict
+    """
+    assert not hasattr(module, 'qconfig'), 'Original module should not have \
+        qconfig object already'
+    module.qconfig = None
+    if prefix in qconfig_dict:
+        module.qconfig = qconfig_dict[prefix]
+    else:
+        module.qconfig = qconfig_parent
+
+    # Insert observers only for leaf nodes, note that this observer is for
+    # the output of the module, for input QuantStub will observe them
+    if module.qconfig is not None and len(module._modules) == 0:
+        # observer will be when we swap the module
+        module.add_module('observer', module.qconfig.activation())
+        # when do we remove hooks?
+        module.observer_hook_handle = module.register_forward_hook(_forward_hook)
+
+    for name, child in module.named_children():
+        if name != 'observer':
+            module_prefix = prefix + '.' + name if prefix else name
+            add_observer(child, qconfig_dict, module.qconfig, module_prefix)
+
+def add_quant_dequant_call(base_fn):
+    r"""Insert calls to quant and dequant around base_fn calls
+
+    Given a `base_fn` function, insert a call to `quant` before the function
+    call which is supposed to quantize the input and a call to `dequant` after
+    the function call which will dequantize the output of the function.
+
+    Args:
+        base_fn: base function we want to call
+
+    Returns:
+        new_fn: a modified function which calls quant -> base_bn -> dequant
+    """
+    def new_fn(x):
+        x = base_fn.__self__.quant(x)
+        x = base_fn(x)
+        x = base_fn.__self__.dequant(x)
+        return x
+    return new_fn
+
+class QuantWrapper(nn.Module):
```

Comment:
docstring

also, does it produce a reasonable repr(m)?

Reply:
we will have extra hierarchy in the leaf node.